lib/pytap13.py

```diff
@@ -17,10 +17,13 @@
 # Author: Josef Skladanka <jskladan@redhat.com>
 
 import re
+
 try:
-    from CStringIO import StringIO
-except ImportError:
+    # Python 2
     from StringIO import StringIO
+except ImportError:
+    # Python 3
+    from io import StringIO
```

Comment:
This looks okay for me.

I didn't know all details and had questions. I summarized questions, answers, details and possible implementatios below.

----

The main questions were the following:

1. Why we need `cStringIO` and fallback to `StringIO`? Should not we save this logic on Python 2?
2. Whether we should use `io.StringIO` or `io.BytesIO` on Python 3?
3. Can we use the same `io.StringIO` or `io.BytesIO` on Python 2?

Answers:

1. For performance matters. Does not matter much here (especially on Python 2 -- not our primary target).
2. The former (`io.StringIO`).
3. Seems so, it should be supported on Python 2.6+.

----

A brief list of facts.

Python 2 facts ([source 1][1], [source 2][2], [source 3][3], [source 4][4], [source 5][5]):

1. `cStringIO.StringIO` is the fast implementation of a string buffer with file-like API. It does not accept a unicode string and have other restrictions (they are not interesting for us here). It is stated as 'optional' and there are suggestions to fallback to `StringIO.StringIO` (see sources 3 and 4; don't know more regarding variety across imterpreter implementations, builds and versions).
2. `StringIO.StringIO` (without `c`) is mostly the same, but supports unicode strings. This, however, is declared as less performant alternative. So we prefer `cStringIO`.
3. `io.StringIO` accepts only unicode strings. It does not accept `str`.
4. `io.BytesIO` accepts only binary strings. It does not accept `unicode`.

Python 3 facts:

1. Now there are two buffer implementations with file-like API: `io.BytesIO` for byte strings and `io.StringIO` for unicode string. (And they both supported in Python 2.6+.)

test-run facts (this particular snippet of the code):

1. There is the typo: `CStringIO` should be `cStringIO`. It is hidden due to the fallback to `StringIO`.
2. When working on Python 2 we're not insterested in decoding from UTF-8 to a unicode string representation and back to UTF-8. We have no problem introduced in Python 3: all string literals becomes unicode and so working with byte strings requires extra attention.
3. Moreover, the only exchange between the `pytap13.py` file and all code outside is accepting a string as input in the `parse()` method.
4. As I see on the top commit on your branch (`open(self.tmp_result, 'r')`), we pass `str` (a binary string) as input on Python 2 and, hmm, `str` (a unicode string) as input on Python 3. Okay.
5. We'll anyway require more weird changes in order to work with both binary and unicode strings inputs: at least have two variants of compiled regexps on Python 3. So let's target only binary strings on Python 2 and only unicode strings on Python 3.

Given those facts we can save the `cStringIO` (and fix `C` to `c`)  usage for performance on Python 2 (but why?):

```python
try:
    # Python 2.
    try:
        # Fast implementation (supports only binary strings).
        from cStringIO import StringIO
    except ImportError:
        # Fallback (supports both string types, but we use it for
        # binary strings).
        from StringIO import StringIO
except:
    # Python 3 (unicode strings).
    from io import StringIO
```

Or we can move to `io.BytesIO` and `io.StringIO` (Python 2.6+):

```python
if sys.version_info[0] == 3:
    # Python 3. Accept a unicode string as input.
    from io import StringIO
else:
    # Python 2. Accept a binary string as input.
    from io import BytesIO as StringIO
```

It looks more straightforward. It'll not raise questions 'hey, Python 2 compatible code should prefer `cStringIO`'. It is a bit weird to rename BytesIO to StringIO, though. Alternative:

```python
if sys.version_info[0] == 3:
    # Python 3. Accept a unicode string as input.
    from io import StringIO as StringBuffer
else:
    # Python 2. Accept a binary string as input.
    from io import BytesIO as StringBuffer
```

However it introduces the new term 'StringBuffer' that is not nice.

We can use `six`:

```python
from six.moves import cStringIO
```

It'll choose `cStringIO.StringIO` on Python 2 and `io.StringIO` on Python 3. It is the ideal variant.

Alternative:

```python
from six import StringIO
```

It'll choose `StringIO.StringIO` on Python 2 (acceptable) and `io.StringIO` on Python 3 (ok).

----

BTW, interesting thread about a performance problem on Python 3 with `BytesIO`, which was resolved in Python 3.5: https://mail.python.org/pipermail/python-dev/2014-July/135542.html. Is it backported to Python 2? Don't know.

----

To sum up, what I would recommend as the general Python 2 & 3 solution in the (quite common) situation, when you need to process binary strings on Python 2 and unicode strings on Python 3? Use `from six.moves import cStringIO` where possible. If `six` usage is discouraged, use the three-cases snippet above. If performance on Python 2 does not matter, use anything.

If performance on Python 3 is matter, then I don't know. Is it okay to operate on unicode strings so? Maybe it worth to do refactoring and use `BytesIO`?

[1]: https://docs.python.org/2/library/stringio.html
[2]: https://docs.python.org/2/library/io.html#io.StringIO
[3]: https://books.google.ru/books?id=PIlQzfnpn5cC
[4]: http://histo.ucsf.edu/BMS270/diveintopython3-r802.pdf
[5]: https://docs.python.org/2/library/io.html#io.BytesIO

Reply:
Reverted block with imports in `lib/pytap13.py`, no sense to change it.